Improve performance of _findUnit

flaviut opened the pull request:
This fixes some performance issues in my heavily-unit-parsing app.

Another idea might be to do an index of reversed unit names, and search in that, but this is much easier to implement and still should provide an improvement in the majority of cases (since I'd imagine that most users tend to prefer a few units at a time, depending on their application).

josdejong commented:
Thanks Flaviu! Your PR looks good at first sight 👍  nice idea to add memoization to `_findUnit`.

Some feedbacks:

1. you're talking about performance improvements. Do you have any feel of how much faster this solution is in practice? (there is a `benchmark` folder where you could add a small benchmark to try out). Since we're adding code, it should be worth it :)
2. It is possible to create new units with `createUnit`. I think that the cached results of `_findUnit` should be reset when there is a new unit created.
3. I see a unit test is still failing, can you look into that?

flaviut commented:
> you're talking about performance improvements. Do you have any feel of how much faster this solution is in practice? (there is a benchmark folder where you could add a small benchmark to try out). Since we're adding code, it should be worth it :)

Great idea! I've added a benchmark, and here are my results:

- before:
  ```
  Unit.parse x 114,313 ops/sec ±0.30% (90 runs sampled)
  evaluate x 4,771 ops/sec ±0.50% (91 runs sampled)
  ```
- after:
  ```
  Unit.parse x 763,288 ops/sec ±0.19% (90 runs sampled)
  evaluate x 17,094 ops/sec ±0.80% (89 runs sampled)
  ```

6.6x improvement in unit parsing time, and 3.6x improvement in evaluating `[1mm, 2mm, 3mm, 4mm, 5mm, 6mm, 7mm, 8mm, 9mm, 10mm]`. Improvements for most people likely wouldn't be this dramatic, I don't know how much unit processing is typical.

> It is possible to create new units with createUnit. I think that the cached results of _findUnit should be reset when there is a new unit created.

Yes! I didn't realize that, but I've added a test and fixed that issue.

> I see a unit test is still failing, can you look into that?

Yup, fixed now. I'd switched to using Map() to back the LRU cache, but forgot to update the tests.

josdejong commented:
wow, that is a huge performance improvement! That's definitely worth it 😃 👍 , very nice!

Thanks for your last updates Flaviu, will now merge your PR.

josdejong commented:
This improvement is published now in `v8.1.1`. Thanks again!
